lib/Frontend/DiagnosticVerifier.cpp

```diff
@@ -547,46 +550,103 @@ DiagnosticVerifier::Result DiagnosticVerifier::verifyFile(unsigned BufferID) {
+    const char *missedFixitLoc = nullptr;
     // Verify that any expected fix-its are present in the diagnostic.
     for (auto fixit : expected.Fixits) {
       // If we found it, we're ok.
-      if (!checkForFixIt(fixit, FoundDiagnostic, InputFile))
-        IncorrectFixit = fixit.StartLoc;
+      if (!checkForFixIt(fixit, FoundDiagnostic, InputFile)) {
+        missedFixitLoc = fixit.StartLoc;
+        break;
+      }
     }
 
-    bool matchedAllFixIts =
-        expected.Fixits.size() == FoundDiagnostic.FixIts.size();
+    bool isUnexpectedFixitsSeen =
+        expected.Fixits.size() < FoundDiagnostic.FixIts.size();
+
+    auto makeActualFixitsPhrase =
+        [&](ArrayRef<DiagnosticInfo::FixIt> actualFixits,
+            std::string &actualFixitsStr) -> std::string {
+      actualFixitsStr = renderFixits(actualFixits, InputFile);
+
+      std::string str("actual fix-it");
+      if (actualFixits.size() >= 2) {
+        str += "s";
+      }
+      str += " seen: ";
+      str += actualFixitsStr;
+      return str;
+    };
+
+    auto emitFixItsError = [&](const char *location, const std::string &message,
+                               const char *replStartLoc, const char *replEndLoc,
+                               const std::string &replStr) {
+      llvm::SMFixIt fix(llvm::SMRange(SMLoc::getFromPointer(replStartLoc),
+                                      SMLoc::getFromPointer(replEndLoc)),
+                        replStr);
+      addError(location, message, fix);
+    };
 
     // If we have any expected fixits that didn't get matched, then they are
     // wrong.  Replace the failed fixit with what actually happened.
-    if (IncorrectFixit) {
+
+    if (missedFixitLoc) {
+      // If we had an incorrect expected fixit, render it and produce a fixit
+      // of our own.
+
+      assert(!expected.Fixits.empty() &&
+             "some fix-its should be expected here");
+
+      const char *replStartLoc = expected.Fixits.front().StartLoc;
+      const char *replEndLoc = expected.Fixits.back().EndLoc;
+
+      std::string message = "expected fix-it not seen";
+      std::string actualFixits;
+
       if (FoundDiagnostic.FixIts.empty()) {
-        addError(IncorrectFixit, "expected fix-it not seen");
+        if (replStartLoc[-1] == ' ') {
+          replStartLoc--;
+        }
+      } else {
+        message +=
+            "; " + makeActualFixitsPhrase(FoundDiagnostic.FixIts, actualFixits);
+      }
+
+      emitFixItsError(missedFixitLoc, message, replStartLoc, replEndLoc,
+                      actualFixits);
+    } else if (expected.noExtraFixitsMayAppear && isUnexpectedFixitsSeen) {
+      // If unexpected fixit were produced, add a fixit to add them in.
+
+      assert(!FoundDiagnostic.FixIts.empty() &&
+             "some fix-its should be produced here");
+
+      const char *noneStartLoc =
+          expected.ExpectedEnd -
+          (fixitExpectationNoneString.size() + 4) /* length of '{{none}}' */;
```

Comment:
I don't think the verifier currently requires `{{none}}` is at the end. I think it probably makes sense to report an error if it does not. I think we can diagnose this just before we start parsing a fix-it on line 469 if Expected.noExtraFixitsMayAppear is already true.

Reply:
I thought that check is needed too.
I was going to make it another PR.

I implemented it and added test case.